src/main/java/com/aws/greengrass/clientdevices/auth/iot/GreengrassClientFactory.java

```diff
@@ -0,0 +1,135 @@
+    private final LazyCredentialProvider lazyCredentialProvider;
+
+    /**
+     * Construct Greengrass Client Factory.
+     *
+     * @param greengrassServiceClientFactory GG Service Client Factory to retrieve the latest device-config
+     * @param lazyCredentialProvider Credential provider for the client
+     */
+    @Inject
+    public GreengrassClientFactory(GreengrassServiceClientFactory greengrassServiceClientFactory,
+                                   LazyCredentialProvider lazyCredentialProvider) {
+        this.greengrassServiceClientFactory = greengrassServiceClientFactory;
+        this.lazyCredentialProvider = lazyCredentialProvider;
+    }
+
+    /**
+     * Provides a new GG v2 client.
+     * Its upto the caller to maintain the lifecycle of the client.
+     * (TODO: This should ideally exist in Nucleus's GreengrassServiceClientFactory)
+     *
+     * @return GreengrassV2Client
+     */
+    public GreengrassV2Client getGGV2Client()  {
+        DeviceConfiguration deviceConfiguration = getDeviceConfiguration();
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
+
+        String awsRegion = Coerce.toString(deviceConfiguration.getAWSRegion());
+        GreengrassV2ClientBuilder clientBuilder = GreengrassV2Client.builder()
+                .httpClient(httpClient.build())
+                .credentialsProvider(lazyCredentialProvider)
+                .overrideConfiguration(
+                        ClientOverrideConfiguration.builder().retryPolicy(RetryMode.STANDARD).build());
+
+        if (Utils.isEmpty(awsRegion)) {
+            return clientBuilder.build();
+        }
+        clientBuilder.region(Region.of(awsRegion));
+        getEnvStage(deviceConfiguration)
+                .ifPresent(stage -> {
+                    String endpoint = RegionUtils.getGreengrassControlPlaneEndpoint(awsRegion, stage);
+                    clientBuilder.endpointOverride(URI.create(endpoint));
+                });
+
+        return clientBuilder.build();
+    }
+
+    /**
+     * Provides a new GG v2 Data client without implicit retry policy.
+     * Consumer should handle retries when appropriate and maintain the lifecycle of the client.
+     * The Nucleus provided V2DataClient has a built in retry policy which cannot be overridden.
+     *
+     * @return GreengrassV2DataClient
+     */
+    public GreengrassV2DataClient getGGV2DataClient()  {
+        DeviceConfiguration deviceConfiguration = getDeviceConfiguration();
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
+
+        String awsRegion = Coerce.toString(deviceConfiguration.getAWSRegion());
+        GreengrassV2DataClientBuilder clientBuilder = GreengrassV2DataClient.builder()
+                .credentialsProvider(lazyCredentialProvider)
```

Comment:
data client should have an anonymous credential provider, it does not use sigv4, data client uses mutual auth.

Reply:
fixed it, thanks.